code/weapon/weapons.cpp

```diff
@@ -3380,6 +3390,29 @@ void weapon_generate_indexes_for_substitution() {
 
 			memset(wip->weapon_substitution_pattern_names, 0, sizeof(char) * MAX_SUBSTITUTION_PATTERNS * NAME_LENGTH);
 		}
+
+		if (wip->failure_rate > 0.0f) {
+			if (stricmp("none", wip->failure_sub_name) != 0) {
+				wip->failure_sub = weapon_info_lookup(wip->failure_sub_name);
+
+				if (wip->failure_sub == -1) { // invalid sub weapon
+					Warning(LOCATION, "Weapon '%s' requests substitution with '%s' which does not seem to exist",
+						wip->name, wip->failure_sub_name);
+					wip->failure_rate = 0.0f;
+				}
+
+				if (Weapon_info[wip->failure_sub].subtype != wip->subtype) {
+					// Check to make sure secondaries can't be launched by primaries and vice versa
+					Warning(LOCATION, "Weapon '%s' requests substitution with '%s' which is of a different subtype.",
+						wip->name, wip->failure_sub_name);
+					wip->num_substitution_patterns = 0;
```

Comment:
Why do you reset `num_substitution_patterns` here? At first glance this has nothing to do with the failure rate feature.